services/heartbeat/etcd/store.go

```diff
@@ -101,6 +158,127 @@ func (c *client) Get(service string) ([]string, error) {
 	return r, nil
 }
 
+func (c *client) Watch(service string) (xwatch.Watch, error) {
+	c.Lock()
+	watchable, ok := c.watchables[service]
+
+	if !ok {
+		watchChan := c.watchChan(service)
+		c.m.etcdWatchCreate.Inc(1)
+
+		watchable = xwatch.NewWatchable()
+		c.watchables[service] = watchable
+
+		go func() {
+			ticker := time.Tick(c.opts.WatchChanCheckInterval())
+
+			for {
+				select {
+				case r, ok := <-watchChan:
+					if ok {
+						c.processNotification(r, watchable, service)
+					} else {
+						c.logger.Warnf("etcd watch channel closed on service %s, recreating a watch channel", service)
+
+						// avoid recreating watch channel too frequently
+						time.Sleep(c.opts.WatchChanResetInterval())
+
+						watchChan = c.watchChan(service)
+						c.m.etcdWatchReset.Inc(1)
+					}
+				case <-ticker:
+					c.RLock()
+					numWatches := watchable.NumWatches()
+					c.RUnlock()
+
+					if numWatches != 0 {
+						// there are still watches on this watchable, do nothing
+						continue
+					}
+
+					if cleanedUp := c.tryCleanUp(service); cleanedUp {
+						return
+					}
+				}
+			}
+		}()
+	}
+	c.Unlock()
+	_, w, err := watchable.Watch()
+	return w, err
+}
+
+func (c *client) watchChan(service string) clientv3.WatchChan {
+	return c.watcher.Watch(
+		context.Background(),
+		servicePrefix(service),
+		// WithPrefix so that the watch will receive any changes
+		// from the instances under the service
+		clientv3.WithPrefix(),
+		// periodically (appx every 10 mins) checks for the latest data
+		// with or without any update notification
+		clientv3.WithProgressNotify(),
+		// receive initial notification once the watch channel is created
+		clientv3.WithCreatedNotify(),
+	)
+}
+
+func (c *client) tryCleanUp(service string) bool {
+	c.Lock()
+	defer c.Unlock()
+	watchable, ok := c.watchables[service]
+	if !ok {
+		// not expect this to happen
+		c.logger.Warnf("unexpected: watches on service %s is already cleaned up", service)
+		return true
+	}
+
+	if watchable.NumWatches() != 0 {
+		// a new watch has subscribed to the watchable, do not clean up
+		return false
+	}
+
+	watchable.Close()
+	delete(c.watchables, service)
+	return true
+}
+
+func (c *client) processNotification(r clientv3.WatchResponse, w xwatch.Watchable, service string) {
+	err := r.Err()
+	if err != nil {
+		c.logger.Errorf("received error on watch channel: %v", err)
+		c.m.etcdWatchError.Inc(1)
+	}
+	// we need retry here because if Get() failed on an watch update,
```

Comment:
feels like this function should return an error and it's the caller's responsibility to perform retries